mitodl/mit-open
# search api

# What are the relevant tickets?
https://github.com/mitodl/mit-open/issues/110

# Description (What does it do?)
This pr replaces the current open search endpoint with an api that allows you to 
search based on query params

The allowed search parameters are
- text 
- offset
-  limit 
- sort - The sort paramenter. Can start with '-' to sort in descentding order." Can also refer to a nested value. For example `-runs.start_date`    
-  resource_type - Valid values are ['course', 'program', 'learning_path', 'podcast','podcast_episode']
-  professional - Valid values are: ['true', 'false']
- certification 
- offered_by - Valid values are: ['mitx', 'ocw', 'bootcamps', 'xpro', 'csail', 'professional education', 'sloan executive education', 'schwarzman college of computing', 'center for transportation & logistics']

 - topic 
 -  department 
 - level 
 - resource_content_tags 
 - aggregations - Valid values are: ['resource_type', 'audience', 'certification', 'offered_by', 'platform', 'topic', 'department', 'level', 'resource_content_tags']

You can search by multiple values for each parameter except text, limit, offset and sort

ex platform=ocw,xpro

# How can this be tested?
1) Make sure you have data in both your database and your open search index. If not, run
- docker-compose run web ./manage.py backpopulate_xpro_data
- docker-compose run web ./manage.py recreate_index --all

2) Go to http://localhost:8063/api/v1/learning_resources_search/ and verify that there are search results

3) Here is the documentation for the allowed parameters http://localhost:8063/api/v1/schema/swagger-ui/#/learning_resources_search/learning_resources_search_retrieve
Play around with them and see that you have the expected results. An example query http://localhost:8063/api/v1/learning_resources_search/?resource_type=course,program&offered_by=xpro&aggregations=resource_type

The expected behavior for aggregations is to include all filters except for the aggregation parameter. For example a search where you have "offered_by" as a search filter and ["offered_by" and "resource_type"] as the aggregation the filter will affect the aggregation results for resource type but not offered by.

A known bug is that sort works for fields that are saved as keyword in the index but not those that are saved as text. So sorting by title, for example throws an error.
